core/metrics-core-service/src/main/java/org/hawkular/metrics/core/service/MetricsServiceImpl.java

```diff
@@ -497,29 +498,36 @@ public void setDefaultTTL(int defaultTTL) {
     }
 
     @Override
-    public Observable<Metric<?>> findAllMetrics() {
-        return dataAccess.findAllMetricsInData()
-                .compose(new MetricFromFullDataRowTransformer(defaultTTL))
+    public Observable<MetricId<?>> findAllMetricIdentifiers() {
+        return dataAccess.findAllMetricIdentifiersInData()
+                .compose(new MetricIdentifierFromFullDataRowTransformer(defaultTTL))
                 .distinct();
     }
 
+    public <T> Observable.Transformer<MetricId<T>, Metric<T>> enrichToMetric() {
+        return t -> t
+                .flatMap(id -> dataAccess.findMetricInMetricsIndex(id)
+                        .compose(new MetricsIndexRowTransformer<>(id.getTenantId(), id.getType(), defaultTTL))
+                        .switchIfEmpty(dataAccess.findMetricInData(id) // This only verifies it exists..
+                                .compose(new MetricFromDataRowTransformer<>(id.getTenantId(), id.getType(), defaultTTL))));
+    }
+
     @Override
     public <T> Observable<Metric<T>> findMetric(final MetricId<T> id) {
-        return dataAccess.findMetricInMetricsIndex(id)
-                .compose(new MetricsIndexRowTransformer<>(id.getTenantId(), id.getType(), defaultTTL))
-                .switchIfEmpty(dataAccess.findMetricInData(id)
-                        .compose(new MetricFromDataRowTransformer<>(id.getTenantId(), id.getType(), defaultTTL)));
+        return Observable.just(id)
+                .compose(enrichToMetric());
     }
 
     @SuppressWarnings("unchecked")
     @Override
     public <T> Observable<Metric<T>> findMetrics(String tenantId, MetricType<T> metricType) {
         Observable<Metric<T>> setFromMetricsIndex = null;
-        Observable<Metric<T>> setFromData = dataAccess.findAllMetricsInData()
+        Observable<Metric<T>> setFromData = dataAccess.findAllMetricIdentifiersInData()
                 .doOnError(Throwable::printStackTrace)
                 .filter(row -> tenantId.equals(row.getString(0)))
                 .distinct()
-                .compose(new MetricFromFullDataRowTransformer(defaultTTL))
+                .compose(new MetricIdentifierFromFullDataRowTransformer(defaultTTL))
+                .flatMap(this::findMetric)
```

Comment:
Right, because of the concatWtih + distinct the metrics emitted from the metricsIndex observable will survive. So can we remove the findMetric call on the setFromData observable?